lucene/core/src/java/org/apache/lucene/util/bkd/BKDRadixSelector.java

```diff
@@ -0,0 +1,350 @@
+
+    //If we are on heap then we just sort on heap
+    if (data instanceof HeapPointWriter) {
+      return heapSort((HeapPointWriter) data, left, right, dim, middle, 0, 0);
+    }
+
+    //reset histogram
+    for (int i = 0; i < bytesPerDim; i++) {
+      Arrays.fill(histogram[i], 0);
+    }
+    //find common prefix, it does already set histogram values if needed
+    int commonPrefix = findCommonPrefix(data, from, to, dim);
+
+    //if all equals we just partition the data
+    if (commonPrefix == bytesPerDim) {
+      return partition(data, left, right, from, to, middle, dim, null, commonPrefix - 1, middle);
+    }
+    //let's rock'n'roll
+    return buildHistogramAndPartition(data, left, right, from, to, middle, 0, commonPrefix, dim,0, 0);
+  }
+
+  void checkArgs(int from, int to, int middle) {
+    if (middle < from) {
+      throw new IllegalArgumentException("middle must be >= from");
+    }
+    if (middle >= to) {
+      throw new IllegalArgumentException("middle must be < to");
+    }
+  }
+
+  private int findCommonPrefix(PointWriter data, int from, int to, int dim) throws IOException{
+    //find common prefix
+    byte[] commonPrefix = new byte[bytesPerDim];
+    int commonPrefixPosition = bytesPerDim;
+    try (PointReader reader = data.getReader(from, to)) {
+      reader.next();
+      byte[] packedValue = reader.packedValue();
+      System.arraycopy(packedValue, dim * bytesPerDim, commonPrefix, 0, bytesPerDim);
+      for (int i =from + 1; i< to; i++) {
+        reader.next();
+        packedValue = reader.packedValue();
+        int j = FutureArrays.mismatch(commonPrefix, 0, commonPrefixPosition, packedValue, dim * bytesPerDim, dim * bytesPerDim + commonPrefixPosition);
+        if (j == 0) {
+          return 0;
+        } else if (j != -1) {
+          commonPrefixPosition =j;
+        }
+      }
+    }
+    //build histogram up to the common prefix
+    for (int i=0; i < commonPrefixPosition; i++) {
+      partitionBucket[i] = commonPrefix[i] & 0xff;
+      partitionBytes[i] = commonPrefix[i];
+      histogram[i][partitionBucket[i]] = to - from;
+    }
+    return commonPrefixPosition;
+  }
+
+  private byte[] buildHistogramAndPartition(PointWriter data, PointWriter left, PointWriter right, int from, int to, int middle,
+                                            int iteration,  int commonPrefix, int dim, int leftCount, int rightCount) throws IOException {
+    //build histogram at the commonPrefix byte
+    try (PointReader reader = data.getReader(from, to)) {
+      if (iteration == 0) {
+        // we specialise this case
+        reader.buildHistogram(dim * bytesPerDim + commonPrefix, histogram[commonPrefix]);
+      } else {
+        while (reader.next()) {
+          byte[] packedValue = reader.packedValue();
+          if (hasCommonPrefix(packedValue, dim, commonPrefix)) {
+            int bucket = packedValue[dim * bytesPerDim + commonPrefix] & 0xff;
+            histogram[commonPrefix][bucket]++;
```

Comment:
maybe we should also write these values that have the right prefix to a temporary OfflinePointWriter so that the next iteration has less data to look at?